### PR TITLE
Fix body validator different error codes for bad Content-Types.

### LINKED ISSUE #628: Request validator passes www-multipart-data when schema is $ref'ed
### Description

Request validator passes a request with Content-type: application/x-www-form-urlencoded when spec only accepts JSON and schema is $ref'erenced.

Interestingly, when schema is not $ref'ed, validation fails.

### Expected behaviour

Validator should reject such request, it contains bad content-type.

### Actual behaviour

Validator passes execution to application code, mayhem ensues.

### Steps to reproduce

Minimal files to reproduce:

#### Swagger spec:

```
swagger: '2.0'

info:
  description: a
  version: "0"
  title: a
    
host: localhost

paths:
            
  /test:
    post:
      operationId: api.test
      consumes:
      - application/json
      parameters:
        - name: body
          description: Just a testing parameter in the body
          in: body
          required: true
          schema:
            type: object
            properties:
              body1:
                type: string
              body2:
                type: string
      responses:
        200:
          description: OK
          
  /test2:
    post:
      operationId: api.test2
      consumes:
      - application/json
      parameters:
        - name: $body
          description: Just a testing parameter in the body
          in: body
          required: true
          schema:
            $ref: '#/definitions/Test'
            
      responses:
        200:
          description: OK
      

definitions:

  Test:
    properties:
      body1:
        type: string
      body2:
        type: string

```

#### API code:
```
#!/usr/bin/env python3

import flask
from flask import request, Response

import connexion
    
def test(body):
    print('body is:', body)
    return "OK"
    
def test2(body):
    print('body is:', body)
    return "OK"


app = connexion.FlaskApp(__name__)
app.add_api('api.yaml', validate_responses=True)

app.run(port=8080, debug=True)
```

#### Test code:
```
#!/usr/bin/env python3

import requests

r = requests.post( "http://localhost:8080/test", headers={'Content-Type':'application/x-www-form-urlencoded'}, data="a=1" )
print( r.status_code ) ## returns 4xx error, somewhat correctly

r = requests.post( "http://localhost:8080/test2", headers={'Content-Type':'application/x-www-form-urlencoded'}, data="a=1" )
print( r.status_code ) ## returns 200

```

#### Requirements:
```
flask
connexion
requests
```

### Additional info:

Output of the commands:

- `python --version`  Python 3.5.2
- `pip show connexion | grep "^Version\:"`  Version: 1.4.2

### PR DESCRIPTION
**Pull request by positron96**
Fixes #628 .

Changes proposed in this pull request:

 - Added a test for this bug.
 - Fixed it by checking for non-empty HTTP POST payload by considering request.body, request.form and request.files (only request.body was checked)

### CODE REVIEW

**Comment by dtkav** [comment]
nice, I noticed this too when working OpenAPI3 support.
Since this may break existing clients, it might make sense to stage it for the 2.0 release.
Just my 2c though.

**Comment by dtkav** [comment]
my bad @positron96 , I think i am confusing it with a different issue where connexion accepts form data even if the mimetype is wrong.

**Comment by positron96** [comment]
That sounds serious. Is there a description of the bug?

**Comment by dtkav** [comment]
I dug into this a little bit to refresh my memory. I couldn't remember if it was a problem with my branch, or also on master.
It turns out that form parameters won't be rejected based on the "consumes" part of the operation.
For example, the spec below still accepts the form parameters, even though `consumes = ["application/json"]`.
```yaml
  /test-formData-param:
    post:
      summary: Test formData parameter
      operationId: fakeapi.hello.test_formdata_param
      consumes:
        - application/json  # connexion assumes application/json if this is blank
      # - application/x-www-form-urlencoded
      parameters:
        - name: formData
          type: string
          in: formData
          required: true
      responses:
        200:
          description: OK
```
The client still sends the correct content type alongside the form data, but the spec doesn't specify that the endpoint consumes form content types, and connexion isn't strict - it let's the form data through.

It's an oddity of the swagger 2.0 spec that you can specify `in: formData` separately from the `consumes` block. This is fixed in the 3.0 spec with the introduction of `requestBody`. The existing behavior may actually make sense for the 2.0 spec - I'm not sure. If you were to reject formData parameter if the `consumes` block didn't match, that should/would probably be in a spec validator. What do you think?

You can reproduce this by running:
```bash
cd tests
PYTHONPATH=$PWD connexion run fixturesimple/swagger.yaml -p 9090
# in another terminal...
curl -X POST -F "formData=somedata" http://localhost:9090/v1.0/test-formData-param
```

Either way, your change looks good to me!

**Comment by positron96** [comment]
@dtkav Yes, I've reproduced this behavior. It seems that Swagger 2.0 vagueness in this point caused it since it's not clear what to do in this case. Probably keeping formData is OK.

Regarding MIME types, #593 is more serious for me, I'll look into it sometime.

File: connexion/decorators/validation.py [hunk 1]
```diff
@@ -101,7 +101,8 @@ def wrapper(request):
             if all_json(self.consumes):
                 data = request.json
 
-                if data is None and len(request.body) > 0 and not self.is_null_value_valid:
+                empty_body = len(request.body) == 0 and len(request.form) == 0 and len(request.files) == 0
```
Comment (jmcs) [inline connexion/decorators/validation.py]: I think `not(request.body or request.form or request.files)` would be more readable.

**Comment by jmcs** [comment]
@positron96 I agree with @dtkav this should target the `dev-2.0` branch (and be included in the changes in the README).

**Comment by dtkav** [comment]
@jmcs It turns out I was conflating this change with some other behavior (elaborated on above).
FWIW I now think this change is relatively benign.

File: connexion/decorators/validation.py (hunk 1 above)
Comment (positron96) [reply connexion/decorators/validation.py]: yes, it definitely will

**Comment by positron96** [comment]
BTW, I've changed the code as requested.

**Pull request merged**